### ISSUE TITLE
Security: Rogue `https://pypi.org/project/chainlit2/` phishing package on PyPI 

### ISSUE THREAD

**Issue by titusz** [Chainlit/chainlit#1281]
I am not sure, but [this](https://pypi.org/project/chainlit2) looks like an impersonating `chainlit2` package on PYPI probably including some malicious code?

**Comment by stephenrs** [comment]
Wow...and this kind of thing is going to continue. Folks will be hunting for exploits in CL, targeting deployments that expose ports for CL...so anyone who is planning to host CL exposed to the internet the way it has currently been prescribed should be strongly warned.

If the CL team cannot guarantee that people won't be able to install root kits through CL, for example, they should recommend (and facilitate) that CL is hosted behind a production-caliber security wall, not as an internet-exposed standalone server/app.

**Comment by dokterbob** [comment]
### Show me those root kits
In the interest of the community!

> Wow...and this kind of thing is going to continue. Folks will be hunting for exploits in CL, targeting deployments that expose ports for CL...so anyone who is planning to host CL exposed to the internet the way it has currently been prescribed should be strongly warned.
> 
> If the CL team cannot guarantee that people won't be able to install root kits through CL, for example, they should recommend (and facilitate) that CL is hosted behind a production-caliber security wall, not as an internet-exposed standalone server/app.

That's a strong statement @stephenrs! We're definitely keeping security in mind and would love to hear from you on such exploits _as part of this package_.

We do not control PyPI and although it is definitely worrying (as well as a sign of success) if people try to make similarly named packages, this is most certainly not something within our control! 

### Proof of maliciousness
Similarly, @titusz, do you have any reason to believe that indeed this alternative build of chainlit is malicious and not just an impatient user who (at the time) wanted to share the latest & greatest (and possibly more unstable) version of chainlit with the world? 

This is something which historically happened a lot on PyPI. So as such, simply having a similarly named package doesn't make it suspect!

Specifically, if we want to get this package pulled from PyPI, we need _**actual proof**_ that the package is malicious. If you have any, please provide it and reopen this ticket and/or contact the maintainers of PyPI directly to act on any potential security issues.

### Installing the latest and greatest from GIT
On the latter notice, I really do hope to allow installation directly from GIT (currently blocked by #1224), so such will no longer be necessary.

**Comment by stephenrs** [comment]
I have to say I'm a bit surprised by the dismissiveness of this response.

First, the chainlit2 project is using your name, branding, and has plagiarized your text. So, unless you have sanctioned the project, whether it contains malicious code or not, it is fraudulent and therefore suspicious, so deserves your attention. It is obviously intended to trick people into thinking the code came from you. Most people wouldn't ignore this.

It also doesn't discuss any distinction between chainlit and chainlit2, so the natural question is what's the difference between the two? It seems unlikely that the difference is better features, and seems at least possible if not likely that the difference is something malicious. So, I'm not sure that simply dismissing this without investigating it is the best course of action for the health/sustainability/reputation of the project/company.

I also want to assume the best about people's intentions, but unfortunately, the real world doesn't always work quite like that.

Also, SSH has been the gold standard of security for as long as any of us can remember, but a few months ago it was announced that an exploit had been discovered in SSH. This motivated anyone who is serious about security to review their sshd server versions and patch/update as needed.

So, we don't need to prove that there are exploits in CL to demonstrate that there should be a concern that it's a possibility. This is the nature of security awareness.

And, since the latest CL doesn't work when put behind an nginx proxy (#1262), it seems the only way to get it to work is to expose it directly to the internet (or localhost). This is inherently dangerous, so my statement isn't strong at all. Until you've had an expert try to penetrate it and fail, then certified it as impenetrable, we simply have no idea how safe it is. So, it's most sensible in a case like this to err on the side of caution. So, my comments stand :)

The perhaps bigger question (maybe for @willydouhard CTO), is: why invite the risk? CL doesn't need to be bulletproof to do the LLM/UI work that it does, so why continue to conceptualize and promote it as "everything you need to run a single-agent dedicated server except the hardware"? It's hard enough to get a new venture going, IMHO you don't need the additional threats highlighted on this issue. Just my 2 cents.

**Comment by stephenrs** [comment]
Also, rather than asking for outsiders to provide proof that the package is malicious, perhaps the responsibility falls to you to download it and prove that it doesn't.

**Comment by dokterbob** [comment]
Thanks for taking the time to write an extensive reply @stephenrs.

Thanks also for pointing out your feedback on #1262, I hadn't noticed that yet and expect to review it today or tomorrow. Indeed, if chainlit doesn't work behind a reverse proxy it should be considered broken, by all standards (including but not limited to security).

Now, neither you nor us are _obliged_ to research _other people's packages_ for security issues. I think it's fair to assume _all_ of use have limited bandwidth and, at this very moment, I am already happy that I managed to address quite a significant number of _internal_ security issues since I started just a few weeks ago. For my own startup, I'm running Chainlit in production -- I can assure you that security is a great _personal_ concern to me.

The above context makes your arguments, which come across a bit like "you're not taking security seriously" come across a bit stingy, hence my somewhat defensive response.

Regarding our security policy, I do think we are soon approaching the point where we need to make clear and public statements on:

1. How issues can/should be reported.
2. What support we do and do not provide, as maintainers _as well as_ **a community**.

With the latter, I would like to clarify that we rely, in large part, on the developer community, on contributors, to push this project forward. This definitely goes for security as well, although Chainlit might (or might not) be interested in offering commercial support, please reach out if you do so desire. 

It is my (_personal_) hope and desire that we'll manage to graduate to a true community effort, as the interest of any single commercial entity to build and maintain an open source package is limited and, as such, not sustainable.

In other words, an interpretation of open source as 'free stuff other people build and support for you, at their expense', is not what this is.

As such, and in general, we do not accept vulnerability reports relating to general plausibilities 'this might be broken'. Or 'it looks like it's broken'. While it is an excellent starting point for security research, I guarantee you won't get awarded a CVE for it. In addition, to my knowledge we are not able to apply dedicated human resources toward security, but if you are interested to provide a budget, perhaps we can talk about involving pentesters, commercially.

Regarding brand infringement etc., this is not with me -- but obviously, if it will be required in order to take a malicious package down, so be it.

**Comment by dokterbob** [comment]
As I already spent so much time on my clarifications here, I just bothered to do a diff between chainlit2 and e1b2dfa4c93ef2f7f6d9f95bdb342f7b7a72a4d7 of the repo and I can confirm that there are 0 differences there.

As such, we won't take action.

In the future, for us to be able to allocate resources more thoughtfully, I would like to ask that any security vulnerabilities are reported based on _actual evidence_ of _specific_ vulnerabilities.

**Comment by stephenrs** [comment]
>As I already spent so much time on my clarifications here, I just bothered to do a diff between chainlit2 and https://github.com/Chainlit/chainlit/commit/e1b2dfa4c93ef2f7f6d9f95bdb342f7b7a72a4d7 of the repo and I can confirm that there are 0 differences there.

Good work, and as my mother used to say "that's what you should have done in the first place" :) But more seriously, I agree with you that it's arguably not your obligation to investigate things like this, so that was a poor choice of words on my part. But, as I said in my follow-up message, it's in the interests of the project to handle potential problems like this, especially when it's so easy to do as in this case. 

I also wouldn't characterize this as a "security vulnerability" in the typical sense because it's not a specific code/system problem, but this kind of thing can be problematic for the health of a project or company more generally. So, I'd think that you would be eager to make sure it was taken care of to protect the interests of the project (and deliver clarity to the community). We may have different approaches and philosophies, however.

>In other words, an interpretation of open source as 'free stuff other people build and support for you, at their expense', is not what this is.

I find this interesting because this project is not an open source project in the typical or traditional sense because a commercial entity is behind it that stands to benefit directly (and perhaps most) from any contributions made by the community. So, I'm not actually sure how closely this aligns with the spirit of "open source" in the truest sense. I'm happy to hear some clarification about what CL/LAI's role is, however.

For anyone who has never read it, I recommend the GNU Manifesto, it's the reason OSS exists: https://www.gnu.org/gnu/manifesto.en.html
